src/services/query.ts

```diff
@@ -53,72 +46,92 @@ export class TimeoutError extends Error {
 export class ServerError extends Error {}
 
 export class QueryTermsService {
-  protected logger: Pino.Logger;
-  protected distribution: Distribution;
-  protected query: string;
-  protected engine: ActorInitSparql;
-  protected timeoutMs: number;
+  private readonly logger: Pino.Logger;
+  private readonly engine: ActorInitSparql;
 
   constructor(options: ConstructorOptions) {
     const args = Joi.attempt(options, schemaConstructor);
     this.logger = args.logger;
-    this.distribution = args.distribution;
-    this.query = args.query;
     this.engine = args.comunica;
-    this.timeoutMs = args.timeoutMs;
   }
 
   // eslint-disable-next-line @typescript-eslint/no-explicit-any
-  protected getConfig(): any {
+  protected getConfig(distribution: Distribution, bindings?: Bindings): any {
     const logger = new LoggerPino({logger: this.logger});
     return {
       log: logger,
       sources: [
         {
           type: 'sparql', // Only supported type for now
-          value: this.distribution.endpoint.toString(),
+          value: distribution.endpoint.toString(),
         },
       ],
-      initialBindings: Bindings({
-        '?query': literal(this.query),
-      }),
+      initialBindings: bindings,
     };
   }
 
-  async run(): Promise<TermsResult> {
-    this.logger.info(
-      `Querying "${this.distribution.endpoint}" with "${this.query}"...`
+  async search(
+    searchQuery: string,
+    distribution: Distribution,
+    timeoutMs: number
+  ) {
+    return this.run(
+      distribution.query,
+      distribution,
+      timeoutMs,
+      Bindings({'?query': literal(searchQuery)})
+    );
+  }
+
+  async lookup(iris: IRI[], distribution: Distribution, timeoutMs: number) {
+    return this.run(
+      // TODO: insert real query here
+      'CONSTRUCT { ?s ?p ?o } WHERE { ?s ?p ?o . VALUES ?s { ?uris } }'.replace(
```

Comment:
Didn’t find a way to use Comunica’s initialBindings in a `VALUES { ?initialBindings }` clause, so doing the replace myself.

Reply:
👍🏻